### ISSUE TITLE
OBSController doesnt stop when connection fails

### ISSUE THREAD

**Issue by rnayabed** [obs-websocket-community-projects/obs-websocket-java#34]
```

OBSRemoteController obsRemoteController = new OBSRemoteController(url, false, pass);


if(obsRemoteController.isFailed())
{
  new StreamPiAlert("Unable to Connect to OBS", "Is it even running? Make sure the websocket plugin is installed. Check Credentials too", StreamPiAlertType.ERROR).show();
}
```

when it fails, the OBSRemoteController keeps on running, whenever you quit the application, the app doesnt stop.

**Comment by TinaTiel** [comment]
The isFailed() method checks the status of a latch object, it doesn't handle any kind of failure. 

I think the way it's meant to be worked with is to register callbacks on different events. Here is an example of how I do this in my own ObsClientManager. I'm using CompleteableFuture as a dirty way of signaling state between threads, but also so that my start, stop, and reload methods are actually blocking on the results of the connection/callback.

```
/*
 * Copyright (c) 2021 TinaTiel. This file is part of the OBS Chatbot project which is released under
 * GNU General Public License v3.0. See LICENSE or go to https://fsf.org/ for more details.
 */

package com.tinatiel.obschatbot.core.client.obs;

import com.tinatiel.obschatbot.core.error.ClientNotAvailableException;
import com.tinatiel.obschatbot.core.request.queue.ActionCommand;
import net.twasi.obsremotejava.OBSRemoteController;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class ObsClientManagerImpl implements ObsClientManager {

    private final Logger log = LoggerFactory.getLogger(this.getClass());
    private final ObsSettings settings;

    CompletableFuture<Void> connected;
    CompletableFuture<Void> disconnected;

    private ObsActionCommandConsumer consumer;
    private OBSRemoteController obsRemoteController;

    public ObsClientManagerImpl(ObsSettings settings) {
        this.settings = settings;
    }

   // snip

    @Override
    public void start() {
        if(obsRemoteController == null) {
            // Initialize the client and shared data
            connected = new CompletableFuture<>();
            disconnected = new CompletableFuture<>();
            obsRemoteController = new OBSRemoteController(
                    "ws://" + settings.getHost() + ":" + settings.getPort(),
                    false,
                    settings.getPassword(),
                    false
            );

            // Define callbacks, we want to complete the future on connection
            obsRemoteController.registerOnError((message, throwable) -> {
                log.error("OBS Client encountered an error: " + message, throwable);
            });
            obsRemoteController.registerCloseCallback((statusCode, reason) -> {
                log.info(String.format("OBS Client closed with statusCode %s and reason '%s", statusCode, reason));
                obsRemoteController = null;
                disconnected.complete(null);
            });
            obsRemoteController.registerConnectionFailedCallback(message -> {
                log.error("Connection to OBS failed: " + message);
            });
            obsRemoteController.registerConnectCallback(response -> {
                connected.complete(null);
            });

            // Connect
            obsRemoteController.connect();

            // Await connection completion, waiting up to the timeout
            try {
                connected.get(settings.getConnectionTimeoutMs(), TimeUnit.MILLISECONDS);
            } catch (InterruptedException interruptedException) {
                log.error("Connection interrupted", interruptedException);
            } catch (ExecutionException executionException) {
                throw new ClientNotAvailableException("Execution failed during start", executionException);
            } catch (TimeoutException e) {
                throw new ClientNotAvailableException("OBS Client failed to start within timeout (" + settings.getConnectionTimeoutMs() + "ms)", e);
            }
            consumer = new ObsActionCommandConsumer(obsRemoteController);
        }
    }

    @Override
    public void stop() {
        log.info("Stopping OBS Client");
        if(obsRemoteController == null) throw new ClientNotAvailableException("No client was available to stop");
        obsRemoteController.disconnect();
        try {
            disconnected.get(settings.getConnectionTimeoutMs(), TimeUnit.MILLISECONDS);
        } catch (InterruptedException interruptedException) {
            log.error("Disconnection interrupted", interruptedException);
        } catch (ExecutionException executionException) {
            throw new ClientNotAvailableException("Execution failed during stop", executionException);
        } catch (TimeoutException e) {
            throw new ClientNotAvailableException("OBS Client failed to stop within timeout (" + settings.getConnectionTimeoutMs() + "ms)", e);
        } finally {
            // Always dereference the controller so it gets garbage-collected
            obsRemoteController = null;
        }
    }

    @Override
    public void reload() {
        log.info("Reloading OBS Client");
        stop();
        start();
    }

}
```

https://github.com/TinaTiel/obs-chatbot/blob/master/core/src/main/java/com/tinatiel/obschatbot/core/client/obs/ObsClientManagerImpl.java

**Comment by TinaTiel** [comment]
Sorry, disregard my last comment -- 

You're right, this could be improved. The connect() method does make a call to onError and onConnectionFailed callbacks if there are problems during the connection process. 

As a workaround, you could register code on those callbacks that calls the disconnect() method if that's the behavior you want. 

But I'm not entirely sure if @LarsBaertschi  etc. may have intended the bot to automatically disconnect/shut down after a failed connection attempt.

**Comment by ChristopheCVB** [comment]
@TinaTiel SHould we close this issue or is it still relevant for 5.x.x?

**Comment by TinaTiel** [comment]
Working on the fix for this today, will be available in 2.0.0
